### PR TITLE
feat(core): reuse remote jwk set and oidc config

### PR DESCRIPTION
**Pull request by IceHe**
## Summary

feat(browser): reuse remote jwks and odic config

## Linear Issue Reference

LOG-1909

## Testing

Pass all existing tests and

image

---
@logto-io/eng

### CODE REVIEW

**Review by demonzoo — APPROVED** [review]
LGTM

**Review by gao-sun — COMMENTED** [review]
unit tests?

File: packages/browser/src/index.ts [hunk 1]
```diff
@@ -335,21 +337,39 @@ export default class LogtoClient {
...
+    const { endpoint } = this.logtoConfig;
+    const discoveryEndpoint = getDiscoveryEndpoint(endpoint);
+    this.oidcConfigPromise = fetchOidcConfig(discoveryEndpoint, this.requester);
+    this.oidcConfig = await this.oidcConfigPromise;
+
     return this.oidcConfig;
   }
 
+  private async getJwtVerifyGetKey(): Promise<JWTVerifyGetKey> {
+    const { jwksUri } = await this.getOidcConfig();
```
Comment (IceHe) [inline packages/browser/src/index.ts:357]: The code here is just for testing.

IF
```ts
const { jwksUri } = await this.getOidcConfig();
```
is not moved above
```ts
if (!this.jwtVerifyGetKey) {
```
`createRemoteJWKSet` will be called 2 times.

---

It's better to rewrite `getJwtVerifyGetKey` like `getOidcConfig` above.

File: packages/browser/src/index.test.ts [hunk 2]
```diff
@@ -39,17 +39,19 @@ const requester = jest.fn();
 const failingRequester = jest.fn().mockRejectedValue(new Error('Failed!'));
 const currentUnixTimeStamp = Date.now() / 1000;
 
+const callInnerFetchOidcConfig = jest.fn(async () => ({
```
Comment (demonzoo) [inline packages/browser/src/index.test.ts:42]: nit: just `fetchOidcConfig` would be fine

File: packages/browser/src/index.ts [hunk 3]
```diff
@@ -335,21 +337,39 @@ export default class LogtoClient {
...
 
+    if (this.oidcConfigPromise) {
+      return this.oidcConfigPromise;
+    }
+
+    const { endpoint } = this.logtoConfig;
+    const discoveryEndpoint = getDiscoveryEndpoint(endpoint);
+    this.oidcConfigPromise = fetchOidcConfig(discoveryEndpoint, this.requester);
+    this.oidcConfig = await this.oidcConfigPromise;
+
```
Comment (demonzoo) [inline packages/browser/src/index.ts:352]: need to remove the cached promise after `await`

**Comment by wangsijie** [comment]
Looks like a "async lock" problem, a simple solution is using lodash.once?

**Comment by IceHe** [comment]
> @wangsijie : Looks like a "async lock" problem, a simple solution is using lodash.once

Use `lodash.once` to refactor.

File: packages/browser/src/index.test.ts [hunk 4]
```diff
@@ -39,17 +39,19 @@ const requester = jest.fn();
...
-  fetchOidcConfig: jest.fn(async () => ({
-    authorizationEndpoint,
-    userinfoEndpoint,
-    tokenEndpoint,
-    endSessionEndpoint,
-    revocationEndpoint,
-    jwksUri,
-    issuer,
-  })),
+  fetchOidcConfig: async () => fetchOidcConfig(),
```
Comment (gao-sun) [inline packages/browser/src/index.test.ts:54]: should use `setTimeout` to simulate network?

File: packages/browser/src/index.ts [hunk 5]
```diff
@@ -197,7 +198,7 @@ export default class LogtoClient {
   }
 
   public async fetchUserInfo(): Promise<UserInfoResponse> {
-    const { userinfoEndpoint } = await this.getOidcConfig();
+    const { userinfoEndpoint } = await this.getOidcConfigOnce();
```
Comment (gao-sun) [inline packages/browser/src/index.ts:201]: it's ok to remove `Once` since the caller doesn't care?

File: packages/browser/src/index.test.ts (hunk 4 above)
Comment (IceHe) [reply packages/browser/src/index.test.ts:54]: Updated. PTAL.

File: packages/browser/src/index.ts (hunk 5 above)
Comment (IceHe) [reply packages/browser/src/index.ts]: Renamed. PTAL.

**Review by gao-sun — APPROVED** [review]

**Pull request merged**